### ISSUE TITLE
Exception when unregistering handlers

### ISSUE THREAD

**Issue by warpech** [StarcounterApps/Website#125]
### Steps to reproduce

(I suppose these are the steps)

1. Run WebsiteEditor, WebsiteProvider, SampleWebsiteTheme
2. Stop WebsiteEditor, WebsiteProvider, SampleWebsiteTheme
3. Run WebsiteEditor, WebsiteProvider, SampleWebsiteTheme
4. Call `http://localhost:8080/website/resetdata` in your browser

### Expected result

Rules created by SampleWebsiteTheme should be unmapped.

### Actual result

I get the following exception:

```txt
Starcounter.DbException: ScErrErrorInHookCallback (SCERR4034): The operation failed because an error was detected in a hook callback function.
Version: 2.3.1.6664.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4034.
   at Starcounter.Query.QueryPreparation.PrepareOrExecuteQuery[T](String query, Boolean slowSql, Object[] values) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Query\QueryPreparation.cs:line 70
   at Starcounter.Query.Sql.SqlEnumCache.CacheOrExecuteEnumerator[T](String query, Boolean slowSQL, Object[] values) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Query\SQL\_SqlEnumCache.cs:line 127
   at Starcounter.Db.FullQueryProcess[T](String query, Boolean slowSQL, Object[] values) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Mock\Db.cs:line 132
   at Starcounter.Db.SlowSQL(String query, Object[] values) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Mock\Db.cs:line 104
   at Website.DataHelper.<>c.<ClearData>b__0_0() in C:\www\Website\src\Website\Helpers\DataHelper.cs:line 14
   at Starcounter.Db.TransactAsync(Action action, UInt32 flags, TransactOptions opts) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Db.cs:line 362
   at Starcounter.Db.Transact(Action action, UInt32 flags, TransactOptions opts) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Db.cs:line 419
   at Website.DataHelper.ClearData() in C:\www\Website\src\Website\Helpers\DataHelper.cs:line 10
   at Website.MainHandlers.<>c.<Register>b__0_2() in C:\www\Website\src\Website\Api\MainHandlers.cs:line 34
   at Starcounter.Rest.UserHandlerInfo.RunUserDelegate(Request req, IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 347
   at Starcounter.Internal.Web.AppRestServer.RunDelegateAndProcessResponse(IntPtr methodSpaceUriSpaceOnStack, IntPtr parametersInfoOnStack, Request req) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter.Apps.JsonPatch\AppRestServer.cs:line 98
   at Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 764
ErrorCode=4034
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4034
---> System.Exception: ScErrHandlerNotFound (SCERR13004): This given handler is not found. Can't unregister HTTP handler "GET /website/blender/surface/u/point/x" on port 8080. Error code: 13004.
Version: 2.3.1.6664.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR13004.
   at Starcounter.GatewayHandlers.UnregisterHttpHandlerInGateway(UInt16 port, String methodSpaceUri) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Apps\GatewayHandlers.cs:line 645
   at Starcounter.Rest.UriHandlersManager.UnregisterHttpHandler(UInt16 port, String methodSpaceUri) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter.Rest\UriHandlersManager.cs:line 931
   at WebsiteProvider.MappingHandlers.UnmapBlendingPoint(WebSection webSection) in C:\www\Website\src\WebsiteProvider\Api\MappingHandlers.cs:line 97
   at WebsiteProvider.CommitHooks.<Register>b__4_2(Object s, WebSection webSection) in C:\www\Website\src\WebsiteProvider\Api\CommitHooks.cs:line 36
   at Starcounter.RuntimeInstalledHookDelegate`1.Invoke(Object triggeringObject) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Hooks\RuntimeInstalledHookDelegate.cs:line 37
   at Starcounter.InvokableHook.InvokeAllWithKey(HookKey key, Object instance) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Hooks\InvokableHook.cs:line 72
   at Starcounter.Db.InvokeOnDelete(IObjectProxy proxy) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Db.cs:line 673
   at Starcounter.Db.InvokeBeforeDeleteHooks(IObjectProxy proxy, UInt64 oid, UInt64 address) in C:\TeamCity\BuildAgent\work\sc-6096\Level1\src\Starcounter\Db.cs:line 616
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR13004

More information about this error may be available in the server error log.
```

##### Involved apps' version

StarcounterApps/Website@12b6887caa63c3bae6652563418c14c38c894085
StarcounterApps/SampleWebsiteTheme@1ad15acbeab09a7251235ce00f92bd91c60f0026

### Screenshot

![2017-06-28 12_57_29-greenshot](https://user-images.githubusercontent.com/566463/27634047-fb0c564c-5c01-11e7-915a-46222c251090.png)

**Comment by un-tone** [comment]
As I understand the reason of the error is the fact that unmapping URI was registered through commit hook from "context" of the SampleWebsiteTheme app with `SelfOnly = true` option and it is not possible to unregister it from another app.
There is a way to fix it fast - https://github.com/StarcounterApps/Website/pull/127/files - in this way the `resetdata` operation works well without errors.
But I'm sure there should be another more correct way to unregister such not own `SelfOnly = true` URI.

@warpech please fix me if I'm wrong.
cc @PatrykSzwer

**Comment by warpech** [comment]
>As I understand the reason of the error is the fact that unmapping URI was registered through commit hook from "context" of the SampleWebsiteTheme app with `SelfOnly = true` option and it is not possible to unregister it from another app.

Can you explain in more details? SampleWebsiteTheme does not map URIs, it only creates objects in the db. SampleWebsiteTheme is not present in the above stack trace. The commit hook is triggered from Website and catched in WebsiteProvider.

So for me it looks like everything you said is right, but it should be Website instead of SampleWebsiteTheme. Or?

**Comment by warpech** [comment]
For me this looks like an error: https://github.com/StarcounterApps/Website/blob/abb16901f032bea08d7b889d7bb60e133b9c56f7/src/WebsiteProvider/Helpers/MappingExtensions.cs#L27-L28

Should be `/websiteprovider/`, not `/website/`.

**Comment by un-tone** [comment]
Yes, I made mistake and was wrong in my last comment.

In this time I looked in the code of Level1 and found out that this error message "Can't unregister HTTP handler" with error code 13004 is thrown from network gateway module ([gateway.cpp:1695](https://github.com/Starcounter/level1/blob/v2.3/src/scnetworkgateway/OurSources/gateway.cpp#L1695)). Also I looked on the implementation of `Handle.UnregisterHttpHandler` method and there is a calling of the handler unregistering from the network gateway ([UriHandlersManager.cs:923](https://github.com/Starcounter/level1/blob/v2.3/src/Starcounter.Rest/UriHandlersManager.cs#L923))

But, the URI from the current issue is registered with enabled `SelfOnly` option and documentation talks us that "`SelfOnly` handlers are not registered in gateway, in comparison with normal handlers." (https://docs.starcounter.io/guides/network/handling-http-requests/)

I see here a bug in the `Handle.UnregisterHttpHandler` implementation.

Another note: [this condition](https://github.com/StarcounterApps/Website/blob/test/Website-125/src/WebsiteProvider/Api/MappingHandlers.cs#L95) in WebsiteProvider returns `true` in the first time, but returns `false` after `Handle.UnregisterHttpHandler` calling with the same parameters. This behavior is correct because unregistering method at this have destroyed handler [UriHandlersManager.cs:921](https://github.com/Starcounter/level1/blob/v2.3/src/Starcounter.Rest/UriHandlersManager.cs#L921)

**Comment by un-tone** [comment]
> Should be `/websiteprovider/`, not `/website/`

@warpech Sure, it should be fixed. In the `develop` branch it have been changed to `/websiteeditor/`, but it is incorrect too.

**Comment by un-tone** [comment]
@alemoi could you please take a look at this

**Comment by alemoi** [comment]
Thanks for finding this! I have added a check for self-only handlers now when unregistering. Should be available in the coming nightly 2.3 and develop.

**Comment by un-tone** [comment]
Checked in SC 2.3.1.6739. It works without errors. Thanks @alemoi !
Closing as fixed.
